### PR TITLE
BUGZ-664: Introduce scaling of rendering for forward renderer

### PR DESCRIPTION
**Pull request by samcake**
https://highfidelity.atlassian.net/browse/BUGZ-664

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20093/HighFidelity-Beta-PR15822-7bf2c5f.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20093/HighFidelity-Quest-Beta-PR15822-7bf2c5f.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20093/HighFidelity-Beta-PR15822-7bf2c5f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20093/HighFidelity-Beta-Interface-PR15822-7bf2c5f.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20095/HighFidelity-Beta-PR15822-a278e6c.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20095/HighFidelity-Quest-Beta-PR15822-a278e6c.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20095/HighFidelity-Beta-PR15822-a278e6c.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20095/HighFidelity-Beta-Interface-PR15822-a278e6c.dmg

**Comment by samcake** [comment]
build this pleaase

**Comment by samcake** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20108/HighFidelity-Beta-PR15822-a278e6c.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20108/HighFidelity-Beta-Interface-PR15822-a278e6c.dmg

File: libraries/render/src/render/ResampleTask.cpp [hunk 1]
```diff
@@ -135,3 +135,50 @@ void Upsample::run(const RenderContextPointer& renderContext, const gpu::Framebu
         args->_viewport = viewport;
     }
 }
+
+gpu::PipelinePointer UpsampleToBlitFramebuffer::_pipeline;
+
+void UpsampleToBlitFramebuffer::configure(const Config& config) {
+    _factor = config.factor;
```
Comment (AndrewMeadows) [inline libraries/render/src/render/ResampleTask.cpp]: Should we be doing any sanity checking on this value?  Are there crash or performance risks if someone configures it completely wrong?

File: libraries/render-utils/src/RenderDeferredTask.cpp [hunk 2]
```diff
@@ -435,6 +428,44 @@ void RenderDeferredTaskDebug::build(JobModel& task, const render::Varying& input
...
+    auto depthFormat = gpu::Element(gpu::SCALAR, gpu::UINT32, gpu::DEPTH_STENCIL); // Depth24_Stencil8 texel format
+    auto primaryDepthTexture = gpu::Texture::createRenderBuffer(depthFormat, frameSize.x, frameSize.y, gpu::Texture::SINGLE_MIP, defaultSampler);
+
+    framebuffer->setDepthStencilBuffer(primaryDepthTexture, depthFormat);
+
+    return framebuffer;
+}
+
+void PreparePrimaryFramebuffer::configure(const Config& config) {
+    _resolutionScale = config.getResolutionScale();
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/RenderDeferredTask.cpp]: Should we be doing any sanity checking on this value?  Are there crash or performance risks if someone configures it completely wrong?

File: libraries/render/src/render/ResampleTask.cpp [hunk 3]
```diff
@@ -135,3 +135,50 @@ void Upsample::run(const RenderContextPointer& renderContext, const gpu::Framebu
...
+
+void UpsampleToBlitFramebuffer::run(const RenderContextPointer& renderContext, const Input& input, gpu::FramebufferPointer& resampledFrameBuffer) {
+    assert(renderContext->args);
+    assert(renderContext->args->hasViewFrustum());
+    RenderArgs* args = renderContext->args;
+    auto sourceFramebuffer = input;
+ //   auto srcFramebuffer = input.get0();
+  //  auto dstFramebuffer = input.get1();
+
+//    resampledFrameBuffer = getResampledFrameBuffer(sourceFramebuffer);
```
Comment (AndrewMeadows) [inline libraries/render/src/render/ResampleTask.cpp]: Are these commented out lines cruft?  Or are they useful?

File: scripts/developer/utilities/lib/prop/PropGroup.qml [hunk 4]
```diff
@@ -110,19 +124,46 @@ PropFolderPanel {
 
     function populateFromObjectProps(object) {
         var propsModel = []
-        var props = Object.keys(object);
 
+        if (Array.isArray(object)) {
+            if (object.length <= 1) {
+            object = object[0];
+            }
```
Comment (AndrewMeadows) [inline scripts/developer/utilities/lib/prop/PropGroup.qml]: Bad indentation at line 130.

File: libraries/render-utils/src/RenderForwardTask.h [hunk 5]
```diff
@@ -17,52 +17,73 @@
...
     void build(JobModel& task, const render::Varying& input, render::Varying& output);
 };
 
 
-class PrepareFramebufferConfig : public render::Job::Config {
+class PreparePrimaryFramebufferMSAAConfig : public render::Job::Config {
     Q_OBJECT
-    Q_PROPERTY(int numSamples WRITE setNumSamples READ getNumSamples NOTIFY dirty)
+    Q_PROPERTY(float resolutionScale  WRITE setResolutionScale READ getResolutionScale)
+    Q_PROPERTY(int numSamples WRITE setNumSamples READ getNumSamples)
```
Comment (SamGondelman) [inline libraries/render-utils/src/RenderForwardTask.h]: should these NOTIFY dirty?

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20108/HighFidelity-Beta-PR15822-a278e6c.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20108/HighFidelity-Beta-Interface-PR15822-a278e6c.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20118/HighFidelity-Beta-PR15822-b0ebcc7.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20118/HighFidelity-Beta-Interface-PR15822-b0ebcc7.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20118/HighFidelity-Beta-PR15822-b0ebcc7.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20118/HighFidelity-Beta-Interface-PR15822-b0ebcc7.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20118/HighFidelity-Beta-PR15822-b0ebcc7.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20118/HighFidelity-Quest-Beta-PR15822-b0ebcc7.apk)

File: libraries/render-utils/src/RenderDeferredTask.cpp (hunk 2 above)
Comment (samcake) [reply libraries/render-utils/src/RenderDeferredTask.cpp]: the sanity check of the _resolucionScale is done in the Config setter
this wouldn't be a problem

File: libraries/render-utils/src/RenderForwardTask.h (hunk 5 above)
Comment (samcake) [reply libraries/render-utils/src/RenderForwardTask.h]: not needed nope

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20124/HighFidelity-Beta-PR15822-8d0b7a8.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20124/HighFidelity-Beta-Interface-PR15822-8d0b7a8.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20124/HighFidelity-Beta-PR15822-8d0b7a8.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20124/HighFidelity-Quest-Beta-PR15822-8d0b7a8.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20124/HighFidelity-Beta-PR15822-8d0b7a8.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20124/HighFidelity-Beta-Interface-PR15822-8d0b7a8.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20130/HighFidelity-Beta-PR15822-962a3c2.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20130/HighFidelity-Beta-Interface-PR15822-962a3c2.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20130/HighFidelity-Beta-PR15822-962a3c2.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20130/HighFidelity-Quest-Beta-PR15822-962a3c2.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20137/HighFidelity-Beta-PR15822-768dd05.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20137/HighFidelity-Beta-Interface-PR15822-768dd05.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20137/HighFidelity-Beta-PR15822-768dd05.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20137/HighFidelity-Quest-Beta-PR15822-768dd05.apk)

**Comment by samcake** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20142/HighFidelity-Beta-PR15822-768dd05.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20142/HighFidelity-Quest-Beta-PR15822-768dd05.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20142/HighFidelity-Beta-PR15822-768dd05.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20142/HighFidelity-Beta-Interface-PR15822-768dd05.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20142/HighFidelity-Beta-PR15822-768dd05.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20142/HighFidelity-Beta-Interface-PR15822-768dd05.exe

**Pull request merged**